jax/_src/api.py

```diff
@@ -1640,6 +1649,8 @@ def f_pmapped(*args, **kwargs):
         global_arg_shapes=tuple(global_arg_shapes_flat))
     return tree_unflatten(out_tree(), out)
 
+  f_pmapped._is_pmap = True
```

Comment:
I think that this is not a pmap-only issue. sharded_jit, xmap and pjit will end up doing the same thing (though e.g. xmap might just crash, because I don't think it works inside a jit).

Reply:
As discussed on chat, perhaps a better solution is that these compiling functions (jit, pmap, sharded_jit, pjit) should return Callable objects which have an `xla_computation(..)` endpoint accepting the same args as the wrapped function but returning an XlaComputation instance equivalent to what they would use when called regularly. I think then they are able to do their own argument handling and the user facing API is trivial:

    f = pmap(f, axis)  # could swap out for jit/sharded_jit/xmap etc
    y = f(x)
    xla_comp = f.xla_computation(x)

@mattjj wdyt?

Reply:
I would definitely prefer that and I think that some folks have actually requested that in the past. Then we could even start deprecating `jax.xla_computation`

Reply:
That sounds like a good idea, though to find out for sure we'd need to work out details. For the short term it might make sense to still land some kind of error check like in this PR (assuming @tomhennigan still wants to pursue it rather than pursuing this new idea).

I agree tagging with `_is_pmap` is likely to be pretty limited/brittle, and not just for the reason cited: it would miss things like `xla_computation(lambda x: pmap(f)(x))`. I think a better implementation mechanism would be just to check the jaxpr that gets produced. After all, it's readily available in this function (i.e. in `computation_maker`). One option is we could check whether it's a single-equation jaxpr, and whether that eqn is the application of a parallel primitive, but maybe there are better versions of this.

Reply:
I'm in favour of implementing the `pmap(f).xla_computation(..)` solution, it feels much better to me to have a solution that doesn't allow you to express the footgun than either warning or raising.

As for whether we need to first make progress on this PR, I think we can drop it. I think xla_computation is obscure enough that the urgency of disarming the footgun is quite low.